internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,281 @@
+	isCached := cachedResp != nil && !cachedResp.Expired && cachedResp.Value != nil
+
+	if isCached {
+		tuplesCacheHitCounter.Inc()
+		span.SetAttributes(attribute.Bool("cached", true))
+		return storage.NewStaticTupleIterator(cachedResp.Value.([]*openfgav1.Tuple)), nil
+	}
+
+	iter, err := dsIterFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &cachedIterator{
+		iter:          iter,
+		tuples:        make([]*openfgav1.Tuple, 0, c.maxResultSize),
+		cacheKey:      cacheKey,
+		cache:         c.cache,
+		maxResultSize: c.maxResultSize,
+		ttl:           c.ttl,
+	}, nil
+}
+
+// Close closes the datastore and cleans up any residual resources.
+func (c *CachedDatastore) Close() {
+	c.OpenFGADatastore.Close()
+}
+
+type cachedIterator struct {
+	iter          storage.TupleIterator
+	tuples        []*openfgav1.Tuple
+	cacheKey      string
+	cache         storage.InMemoryCache[any]
+	ttl           time.Duration
+	maxResultSize int
+	closeOnce     sync.Once
+	wg            sync.WaitGroup
+	sf            singleflight.Group
+}
+
+func (c *cachedIterator) addToBuffer(t *openfgav1.Tuple) bool {
+	if c.tuples == nil {
+		return false
+	}
+	c.tuples = append(c.tuples, t)
+	if len(c.tuples) >= c.maxResultSize {
+		tuplesCacheDiscardCounter.Inc()
+		c.tuples = nil
+	}
+	return true
+}
+
+// Next see [Iterator.Next].
+func (c *cachedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	t, err := c.iter.Next(ctx)
+	if err != nil {
+		if !errors.Is(err, storage.ErrIteratorDone) && !errors.Is(err, context.Canceled) && !errors.Is(err, context.DeadlineExceeded) {
+			c.tuples = nil // don't store results that are incomplete
+		}
+		return nil, err
+	}
+
+	c.addToBuffer(t)
+
+	return t, nil
+}
+
+// Stop see [Iterator.Stop].
+func (c *cachedIterator) Stop() {
+	c.closeOnce.Do(func() {
+		if c.tuples == nil {
+			c.iter.Stop()
+			return
+		}
+		// prevent goroutine if iterator was already consumed
+		ctx := context.Background()
+		if _, err := c.iter.Head(ctx); errors.Is(err, storage.ErrIteratorDone) {
+			c.flush()
+			c.iter.Stop()
+			return
+		}
+
+		c.wg.Add(1)
+		go func() {
+			defer c.iter.Stop()
+			defer c.wg.Done()
+			// prevent draining on the same iterator across multiple requests
+			_, _, _ = c.sf.Do(c.cacheKey, func() (interface{}, error) {
+				for {
+					// attempt to drain the iterator to have it ready for subsequent calls
+					t, err := c.iter.Next(ctx)
+					if err != nil {
+						if errors.Is(err, storage.ErrIteratorDone) {
+							c.flush()
+						}
+						break
+					}
+					// if the size is exceeded we don't add anymore and exit
+					if !c.addToBuffer(t) {
+						break
+					}
+				}
+				return nil, nil
+			})
+		}()
+	})
+}
+
+// Head see [Iterator.Head].
+func (c *cachedIterator) Head(ctx context.Context) (*openfgav1.Tuple, error) {
+	return c.iter.Head(ctx)
+}
+
+func (c *cachedIterator) flush() {
+	if c.tuples == nil {
+		return
+	}
+	tuples := make([]*openfgav1.Tuple, len(c.tuples))
+	copy(tuples, c.tuples)
```

Comment:
Why do we need to do the copy?

Reply:
Want to allow `cachedIterator` to be GC'ed since otherwise the cache will hold the pointer. I could be wrong in this but I think it increases memory and GC backpressure.

Reply:
probably good to add this as comment

Reply:
added comments